docs/hooks/new-language.rst

```diff
@@ -1,105 +1,120 @@
+#. The hooks handler opens a TCP socket on port 61321.
+#. Dredd waits for a moment and then tries to connect to ``http://127.0.0.1:61321``. In case the connection is not successful, it exits with status code 3.
+#. For each :ref:`type of hooks <types-of-hooks>` Dredd creates a message and sends it to the socket. The message contains UUID and serialized :ref:`transaction object <transaction>` (or an array of them, in case of `beforeAll`, `afterAll`). Individual messages are sent as UTF-8 JSON documents separated by a newline.
+#. Hooks handler reads a message, calls a corresponding hook code, and sends back a message with modified contents.
+#. Dredd awaits a message with corresponding UUID. Once it arrives, Dredd overwrites its internal HTTP transaction data with the ones from the incoming message.
 
-   -  `Must return message ``Starting`` to stdout <https://github.com/apiaryio/dredd-hooks-template/blob/master/features/tcp_server.feature#L5>`__
+.. image:: ../_static/images/hooks-handler.png
+   :align: center
 
--  Hooks API in your language for registering code being executed during the :ref:`Dredd lifecycle <execution-life-cycle>`:
 
-   -  before all transactions
-   -  before each transaction
-   -  before transaction
-   -  before each transaction validation
-   -  before transaction validation
-   -  after transaction
-   -  after each transaction
-   -  after all transactions
+Implementation guide
+--------------------
 
--  When CLI command is executed
+A hooks handler is a CLI command, which implements following:
 
-   -  It loads files passed in alphabetical order with paths resolved to absolute form
+-  It accepts paths to hook files as arguments. They are already passed resolved as absolute paths, in the right order.
+-  It allows users to register hook functions in the hook files, i.e. it provides a *hooks API* similar to those in other hooks handler implementations (see :ref:`JavaScript <hooks-nodejs>`, :ref:`Python <hooks-python>`, :ref:`Ruby <hooks-ruby>`). It allows to register :ref:`all types of hooks supported by Dredd <types-of-hooks>`.
+-  It loads the hook files and registers any hook functions found in them for later execution.
+-  It runs a TCP socket server on port ``61321`` and prints ``Starting`` to ``stdout`` when ready.
 
-      -  It exposes API similar to those in :ref:`Ruby <hooks-ruby>`, :ref:`Python <hooks-python>` and :ref:`Node.js <hooks-nodejs>` to each loaded file
-      -  It registers functions declared in files for later execution
 
-   -  starts a TCP socket server and starts listening on ``http://127.0.0.1:61321``.
+Handling hooks
+^^^^^^^^^^^^^^
 
--  When any data is received by the server
+When any data is received by the TCP server, the hooks handler:
 
-   -  Adds every received character to a buffer
-   -  When delimiting newline (``\n``) character is received
+-  Adds every received character to a buffer.
+-  When the delimiter newline character ``\n`` is received:
```

Comment:
Please represent `\n` as hex or something, `\n` is a common pattern on describing this but not all languages use it afaik.